doc/fts_indexer.1
.\" Man page for fts_indexer
.\" Use the following command to view man page:
.\"
.\"  tbl fts_indexer.1 | nroff -man | less
.\"
.TH FTS_INDEXER 1 "December 28, 2013"
.SH NAME
fts_indexer \- A tool to index Emdros databases for Full Text Search
.SH SYNOPSIS
\fBfts_indexer\fR [ options ] [filters]
.br
.SH DESCRIPTION
\fBfts_indexer\fR is a command-line tool to prepare an Emdros database
for Full Text Search using the techniques of the fts_search tool (or
the libharvest library).

The \fBfts_indexer\fR tool needs a source database to read from, an
optional target database to write to (which by default is the same as
the source database), and an optional output file name to (also) dump
the MQL to.

The program also needs a "bookcase object type", which must consist of
objects that are somehow "larger" than the "indexed object type".
That is, the "indexed object type" (for example, "Word") must have
objects which are wholly contained within the "bookcase object type"
(for example, "document").  In effect, the "bookcase object type"
provides the frames of monads within which to index the "indexed
object type".  In addition, the "indexed object type" must have a
concomitant "indexed feature", which is the feature on which to index.
For example, "surface" could be a feature on "Word" which could be
ideal to index on.

The program will issue a series of "CREATE OBJECT TYPE" and "CREATE
OBJECTS WITH OBJECT TYPE" statements on the target database.
Optionally, with the --nodb switch, one can skip the actual creation
of the data in the database.  This is usually used with the -o switch,
which dumps the MQL statements to a file.  It is possible both to dump
the MQL statements to a file, and to create the data in the database
on-the-fly at the same time.


.SH OPTIONS
\fBfts_indexer\fR supports the following command-line switches:
.TP 6
.BI \-\-help
show help
.TP
.BI \-V\ ,\ \-\-version
show version
.TP
.BI \-fe\ ,\ \-\-fts\-engine \ fts\-engine\-version
Use the given version of the FTS engine. Currently, versions 1 and 2
are supported. Version 1 is the current default.
.TP
.BI \-\-nodb
Do not create objects or object types in the target database.  In
fact, do not connect to the target database at all for creation of
objects or object types.
.TP
.BI \-\-no\-vacuum
Do not emit a VACUUM DATABASE ANLYZE GO statement at the end of the
indexing.  This can be useful if more than one invocation is needed on
a particular database, and one intends to run VACUUM DATABASE ANALYZE
only on the last invocation.  This can save some time otherwise
wasted.
.TP
.BI \-d\ ,\ \-\-dbname \ dbname 
set database from which to read, and to which to write (unless -td or
--nodb is given).
.TP
.BI \-td\ ,\ \-\-target-dbname \ dbname 
set target database to which to write.  Normally, the target database
name is the same as the source database, but this switch allows you to
override that behavior and choose another database.  The database
should exist.
.TP
.BI \-\-bookcase\-otn \ object\-type\-name
Use object-type-name as the object type which surrounds the "indexed
object type".
.TP
.BI \-\-indexed\-otn \ object\-type\-name
Use object-type-name as the object type which must be indexed.
.TP
.BI \-\-indexed\-feature \ feature\-name\[, ...]
Use feature-names in the comma-separated list as the feature(s) on the
indexed object type for which to generate the index.  That is, if the
indexed object type is, for example, "Word", then the indexed feature
could be "surface", and then "Word" must have a feature called
"surface".  Alternatively, the feature-name list could be
"surface,stemmed_surface", meaning that the two features "surface" and
"stemmed_surface" would both be indexed in the same index.
.TP
.BI \-o\ ,\ \-\-output \ filename
dump MQL statements to file filename. If this switch is not given,
then the MQL is just executed within the target database.
.TP
.BI \-h\ ,\ \-\-host \ hostname 
set source db back-end hostname to connect to (default
is 'localhost').  Has no effect on SQLite or SQLite 3.  If --nodb is
not given, and if -th or --target-host is not given, then the target
database host will be the same as the source database host.
.TP
.BI \-th\ ,\ \-\-target\-host \ hostname 
set target db back-end hostname to connect to (default
is 'localhost').  Has no effect on SQLite or SQLite 3.  Has no effect
it --nodb is given.  If -th and --target-host are not given, then the
target database host will be the same as the source database host.
.TP
.BI \-u\ ,\ \-\-user \ user
set source database user to connect as (default is 'emdf').  Has no
effect on SQLite or SQLite 3.  Will be used for the target database
user as well, unless --nodb, -tu, or --target-user is given.
.TP
.BI \-tu\ ,\ \-\-target\-user \ user
set target database user to connect as (default is 'emdf').  Has no
effect on SQLite or SQLite 3.  Will override the source database user
if given.  Has no effect if --nodb is given.
.TP
.BI \-p\ ,\ \-\-password \ password
set source database password to use for the source database user.  Has
no effect on SQLite or SQLite 3, unless you have an encryption-enabled
SQLite (3), in which case this gets passed as the key.  Will be used
for the target database password, unless one of --nodb, -tp or
--target-password is given.
.TP
.BI \-tp\ ,\ \-\-target\-password \ password
set target password to use for the target database user.  Has no
effect on SQLite (3), unless you have an encryption-enabled SQLite
(3), in which case this gets passed as the key.  Will override the
source database user for the target database.  Has no effect if --nodb
is given.
.BI \-b\ ,\ \-\-backend \ backend
set source database backend to `backend'. The target database backend
is also given with this switch, unless provided by the -tb switch.
Valid values are: For PostgreSQL: "p", "pg", "postgres", and
"postgresql". For MySQL: "m", "my", and "mysql". For SQLite 2.X.X:
"2", "s", "l", "lt", "sqlite", and "sqlite2". For SQLite 3.X.X: "3",
"s3", "lt3", and "sqlite3".
.TP
.BI \-tb\ ,\ \-\-target\-backend \ backend
set target database backend to `backend'. The target database backend
is usually the same as the source database backend, but this switch
makes it possible to use another database backend for the
target. Valid values are the same as for -b.
.TP
.BI \-sf\ ,\ \-\-stylesheet-filename \ stylesheet-filename
Gives the name of the file containing the JSON stylesheet to use. See
the man-page fts_filters(5) to learn the syntax of that file.

This option must be used in conjunction with the -s option.  If it is
used, you cannot also use filters at the end of the command line.
.TP
.BI \-s\ ,\ \-\-stylesheet \ stylesheet-name
This option tells the program which stylesheet within the stylesheet
file provided with the -sf option must be used.

This option must be used in conjunction with the -sf option.  If it is
used, you cannot also use filters at the end of the command line. 

.SH FILTERS
After the last option, you have the possibility to specify any number
of filters through which the indexed feature string must be passed
before being stored.

Some filters take an argument, in which case the argument must come
directly after the filter name, separated by whitespace on the command
line.

You cannot use both filters on the command line and the -s and -sf
options.

The following filters are available:

.TP 6
.BI strip\-whitespace
Takes no parameter; just strips whitespace from either end of the
string.
.TP
.BI strip\-chars \ chars\-to\-strip
Takes 1 parameter; strips all characters in the parameter from either
end of the string.
.TP
.BI lowercase
Takes no parameter; makes the string lower-case. Only works with ASCII
letters.
.TP
.BI uppercase
Takes no parameter; makes the string upper-case. Only works with ASCII
letters.

For example, the following chain:

strip-whitespace lowercase strip-chars '.,;:?!'

will first strip whitespace from either end, then make all ASCII
letters lower-case, then strip any of the punctuation chars '.,;:?!'
from either end of the string.


.SH RETURN VALUES
.TP
.BR 0 " Success"
.TP
.BR 1 " Wrong usage"
.TP
.BR 2 " Connection to backend server could not be established"
.TP
.BR 3 " An exception occurred (the type is printed on stderr)"
.TP
.BR 4 " Could not open file"
.TP
.BR 5 " Database error"
.TP
.BR 6 " Compiler error (internal error)
.SH AUTHORS
Written Ulrik Sandborg-Petersen (ulrikp@emdros.org).


NEWS
This file is for release-notes for Emdros.

- *** Version 3.7.0 *** 

Release notes for Emdros 3.7.0
==============================

- Emdros has been relicensed under the MIT license.

- All copyright in Emdros and related documentation and build scripts
  has been assigned to Ulrik Sandborg-Petersen's holding company,
  Sandborg-Petersen Holding ApS.  This is so as to be able to serve
  customers better with support and other Emdros-related help.


- *** Version 3.6.1 *** 

Release notes for Emdros 3.6.1
==============================

Emdros version 3.6.1 is a maintenance release with the following
summary of changes since 3.6.0:

- The MQL query language has been extended with the ability to compare
  features with type LIST OF INTEGER, LIST OF ID_D, and LIST OF
  <enum-type> with an explicit list (including the empty list) using
  the "=" comparison operator.

  For example:

  SELECT ALL OBJECTS
  WHERE
  [Clause_atom
     daughters = () // Test for empty list
     OR
     daughters = (12345,67891) // Test for explicit, non-empty list
  ]
  GO

- Detection of PHP 5 and PHP 7 on macOS now works correctly.

- Compileability on Solaris 10 with Sun's CC C++ compiler has been
  restored.

- Compatibility with wxWidget 2.8.12 has been restored.


Enjoy!


Ulrik Sandborg-Petersen



- *** Version 3.6.0 *** 

Release notes for Emdros 3.6.0
==============================

The release notes below cover some of the important changes in Emdros
since 3.5.0.

- SQLite2 support has been removed, as version 2 of SQLite has not
  been updated by the SQLite authors for many, many years.  Also,
  SQLite2 support in Emdros has been been deprecated for some time.

- The MQL subsystem has a better query planner, which is able to
  choose the "Innermost First" strategy, rather than the default
  "Outermost First" strategy, under many more circumstances.

- The Emdros MySQL backend now has support for transactions, if
  available in the MySQL server.

- Assorted performance improvements to the following subsystems:

  - XML parsing (renderobjects, for example)
  - MQL

- Node.JS support has been removed.  The reason is that the latest
  SWIG version (3.0.12) does not support the latest Node.  Also, the
  Node.JS bindings never really worked.

- PCRE has been upgraded.

- SQLite3 has been upgraded.

- JSON output now escapes string characters correctly.

- Compiling Linux .rpm packages works on recent systems (CentOS 7,
  Fedora 28).

- Compiling Linux .deb packages works on recent systems (Ubuntu 16.04,
  Ubuntu 18.04, Debian 9.4).

- Building on Windows supports the latest Visual Studio.

- Lots of compiler warnings have been removed.

- the harvest template language has new constructs:

  - {{ inccounter <string> var <string> }}
  - {{ deccounter <string> var <string> }}

- The Full Text Search FTS3 subsystem is now able to obey
  largest_proximity, and to use * and ? in the query terms.

- A new program was added, bptdetach(1), which enables detaching
  "payloads" from BPT databases.

- The BPT backend (if available -- ask Ulrik for the source code
  licensing terms if you are interested) now supports LZMA compression
  of payloads.

- Source code using wxWidgets has been updated to support wxWidgets
  2.8, 3.0, and 3.1.

- The Emdros PostgreSQL backend now adds "TEMPLATE template0" when
  doing CREATE DATABASE.  This is so as to avoid using template1,
  which may conflict with the encoding used (and does so on Debian
  9.4).

- The amalgamation output has been split into several files.  This is
  because the Android NDK MIPS target cannot compile the old,
  monolithic amalgamation files.



- *** Version 3.5.0 *** 

Release notes for Emdros 3.5.0
==============================

The focus-areas of this release include:

- Enhancements to the core MQL
- Bugfixes to the core MQL.
- Enhancements to the render library
- Enhancements to the Full Text Search subsystem
- Enhancements to SWIG bindings
- Enhancements to the build system

Please consult the ChangeLog (going back to version 3.4.1.pre01) for
the specifics.  You can also consult the MQL Programmer's Guide for
the enhancements to the core MQL, and the Programmer's Reference Guide
for the enhancements to the Render library and the FTS subsystem.

Briefly, the core MQL has been enhanced as follows:

- SELECT ALL OBJECTS has a new WITH MAX RANGE clause, helping to limit
  otherwise runaway queries such as [Word]..[Word].

- SELECT ALL OBJECTS now sports an [UnorderedGroup] block.

- GET AGGREGATE FEAUTURES is a new statement, similar to SQL's MIN(),
  MAX(), SUM(), and COUNT() aggregate functions.


- *** Version 3.4.0 *** 

Release notes for Emdros 3.4.0
==============================

The focus-areas of this release include:

- Support for iOS and Android
- A smoother user experience for the GUI applications
- Enhancements to the MQL query language
- Documentation improvements
- Build improvements on all platforms
- Speed improvements


Binaries for Windows(R) and Mac OS X(R)
---------------------------------------

As usual, binaries are available for Mac OS X and Windows(R).

The Windows binaries are available in both x86 and x64 versions.  They
have built-in support for MySQL, SQLite 2, and SQLite 3.  The x86
binaries are built with Visual Studio Express 2010.  The x64 binaries
are built with Windows SDK 6.0A.

The Mac OS X binaries are Universal binaries (i386 and x86_64) running
on Mac OS X 10.6 (Snow Leopard), 10.7 (Lion), and 10.8 (Mountain
Lion), 10.9 (Mavericks), and possibly later versions of Mac OS X.
They do not have support for either MySQL or PostgreSQL; Only SQLite 2
and SQLite 3 are supported in the Mac OS X binaries.  You might be
able to compile the sources with support for MySQL yourself, and
possibly also PostgreSQL.

Linux binaries

src/fts_indexer.cpp
/*
 * fts_indexer.cpp
 *
 * A front-end to the BookcaseIndexer class in the libharvest library.
 * Created: 5/1-2001 (1st of May, 2001)
 * Last update: 5/11-2018
 *
 * Return codes:
 * 
 * 0 : Success
 * 1 : Wrong usage
 * 2 : Connection to backend server could not be established
 * 3 : An exception occurred
 * 4 : Could not open file
 * 5 : Database error
 * 6 : Compiler error
 *
 */

/*
 * Copyright (C) 2001-2018     Ulrik Sandborg-Petersen
 * Copyright (C) 2018-present  Sandborg-Petersen Holding ApS, Denmark
 *
 * Licensed under the MIT License.
 *
 * Please see the file COPYING in the root of the sources for more details.
 *
 */



#include <emdros-config.h>
// Do this because emdros-config might #define malloc,
// in which case some systems will fail to use the real 
// malloc, thus causing build errors.
#undef malloc


#include <monads.h>
#include <iostream>
#include <sstream>
#include <fstream>
#include <string_func.h>
#include <emdfdb.h>
#include <table.h>
#include <emdros-lconfig.h>
#include <emdf_output.h>
#include <emdros_environment.h>
#include <encryption.h>
#include <opt.h>
#include <messages.h>

#include <harvest_fts.h>
#include <harvest_fts2.h>


// TODO: Add filters
int exec_indexer(EmdrosEnv *pSourceEE, EmdrosEnv *pTargetEE, const std::string& output_filename, const std::string& bookcase_otn, const std::string& indexed_otn, const std::string& indexed_feature, const std::string& stylesheet_name, const std::string& stylesheet_file_name, const StringList& filter_string_list, bool bEmitVacuum, int fts_engine_version)
{ 
	bool bResult = true;

	std::ofstream *pFOUT = 0;
	if (output_filename.empty()) {
		pFOUT = 0;
	} else {
		pFOUT = new std::ofstream(output_filename.c_str());
	}

	if (pFOUT != 0 && !(*pFOUT)) {
		delete pFOUT;
		std::cerr << "ERROR: Could not open file '" << output_filename << "' for writing." << std::endl;
		return 4;
	}

	// from harvest/harvest_fts.cpp.
	EmdrosFTSStringFilter *pFilterChain = 0;

	if (stylesheet_name == "") {
		pFilterChain = stringFilterFactory(filter_string_list); 
	} else {
		std::string filter_error_message;
		pFilterChain = stringFilterFactoryJSONFileName(stylesheet_file_name, bResult, filter_error_message, stylesheet_name);
		if (!bResult) {
			std::cerr << "ERROR doing parsing the filters:\n" 
				  << filter_error_message
				  << std::endl;
		}
	}

	if (bResult) {
		BookcaseIndexer *pIndexer = 0;
		if (fts_engine_version == 1) {
			pIndexer = new BookcaseIndexer(pSourceEE, pTargetEE, pFOUT, bookcase_otn, indexed_otn, indexed_feature, pFilterChain);
		} else if (fts_engine_version == 2) {
			pIndexer = new BookcaseIndexer2(pSourceEE, pTargetEE, pFOUT, bookcase_otn, indexed_otn, indexed_feature, pFilterChain, 50000);
		} else {
			std::cerr << "ERROR: Unknown FTS engine version: " << fts_engine_version << ". Aborting." << std::endl;
			bResult = false;
		}

		if (bResult) {
			std::cout << "About to start indexing..." << std::endl;
			// TODO: Add --fm and --lm options, using them here!
			// Right now, the empty set means 'everything'.
			if (!pIndexer->process(SetOfMonads(), bEmitVacuum)) {
				std::cerr << "ERROR doing the indexing:\n" 
					  << pIndexer->errorMessage()
					  << std::endl;
				bResult = false;
			} else {
				bResult = true;
			}
		}
		delete pIndexer;
	}

	delete pFOUT;
	delete pFilterChain;

	if (bResult) {
		std::cout << "...Finished indexing!" << std::endl;
		return 0;
	} else {
		std::cerr << "...Indexing finished with error!" << std::endl;
		return 5;
	}
}

void show_version(std::ostream& ostr)
{
	ostr << "fts_indexer from Emdros version " << EMDROS_VERSION << std::endl;
}

void print_usage(std::ostream& ostr)
{
	show_version(ostr);
	ostr << "Usage:" << std::endl;
	ostr << "   fts_indexer [options]" << std::endl;
	ostr << "   fts_indexer (--help|-V|--version)" << std::endl;
	ostr << "OPTIONS:" << std::endl;
	printUsageStandardArguments(ostr);
	ostr << "   -fe , --fts-engine (1|2)       Use the given FTS engine version" << std::endl;
	ostr << "   -d , --dbname <dbname>  Set db name to read from, and also to write to,\n"
	     << "                           unless -td is used to indicate where to write." << std::endl;
	ostr << "   -td , --target-dbname <dbname>  Set target database name" << std::endl;
	ostr << "   -th , --target-host <hostname> Set target database host name" << std::endl;
	ostr << "   -tu , --target-user <user> Set target password user name" << std::endl;
	ostr << "   -tp , --target-password <password> Set target database password" << std::endl;
	ostr << "   -tb , --target-backend <backend> Set target database backend." << std::endl;
	ostr << "                        <backend> must be drawn from the same values" << std::endl;
	ostr << "                        as for the -b switch." << std::endl;
	ostr << "   -sf , --stylesheet-filename <stylesheet-filename> The name of the file" << std::endl;
	ostr << "                        from which to read the JSON stylesheet." << std::endl;
	ostr << "   -s , --stylesheet <stylesheet-name> The name of the stylesheet to use" << std::endl;
	ostr << "                        from the stylesheet file." << std::endl;
	ostr << "   --bookcase-otn <otn> Use otn as the object type for the bookcase" << std::endl;
	ostr << "   --indexed-otn <otn>  Use otn as the object type to be indexed" << std::endl;
	ostr << "   --indexed-feature <fn>  Use comma-separated list fn as the feature name(s)\n"
	     << "                           to index on the indexed object type to index." << std::endl;
	ostr << "   --nodb               Do not alter target database" << std::endl;
	ostr << "   --no-vacuum          Do not emit 'VACUUM DATABASE ANALYZE GO' at the end." << std::endl;
	ostr << "   -o , --output <filename>  Output to filename" << std::endl;
	ostr << "DEFAULTS:" << std::endl;
	ostr << "   -fe 1" << std::endl;
	printUsageDefaultsOfStandardArguments(ostr);

}

int main(int argc, char* argv[])
{
	// Set defaults
	bool bDontUseTargetDB = false;
	bool bEmitVacuum = true;
	eOutputKind output_kind = kOKConsole;
	eCharsets charset = kCSUTF8;
	eBackendKind source_backend_kind;
	eBackendKind target_backend_kind;
	std::string output_filename;
	std::string source_db_name;
	std::string target_db_name;
	std::string source_hostname;
	std::string target_hostname;
	std::string source_user;
	std::string target_user;
	std::string source_password;
	std::string target_password;
	std::string bookcase_otn;
	std::string indexed_otn;
	std::string indexed_feature;
	std::string stylesheet_file_name;
	std::string stylesheet_name;
	StringList filter_string_list;
	int fts_engine_version;

	addStandardArguments(false // Do add -e | --encoding
			     );

	addOption("-s", "--stylesheet", true,
		  "",
		  "ERROR: -s and --stylesheet must have a stylesheet name as their argument.\n"
		  "       example: -s 'englishstylesheet'\n");

	addOption("-sf", "--stylesheet-filename", true,
		  "",
		  "ERROR: -sf and --stylesheet-filename must have a filename as their argument.\n"
		  "       example: -sf /home/myuser/stylesheet.json\n");

	addOption("-td", "--target-dbname", true,
		  "",
		  "ERROR: -td and --target-dbname must have a database-name as their argument.\n"
		  "       example: -td 'mydb'\n");

	addOption("-tp", "--target-password", true,
		  "",
		  "ERROR: -tp and --target-password must have a password as their argument.\n"
		  "       example: -tp 'sicret'\n");
	addOption("-th", "--target-host", true,
		  "",
		  "ERROR: -th and --target-host must have a database hostname as their argument.\n"
		  "       example: -tp 'sicret'\n");
	addOption("-tu", "--target-user", true,
		  "",
		  "ERROR: -tu and --target-user must have a database user as their argument.\n"
		  "       example: -tp 'sicret'\n");
	addOption("--bookcase-otn", "--bookcase-otn", true,
		  "",
		  "ERROR: --bookcase-otn must have an object type name as its argument.\n"
		  "       example: --bookcase-otn Document\n");
	addOption("--indexed-otn", "--indexed-otn", true,
		  "",
		  "ERROR: --indexed-otn must have an object type name as its argument.\n"
		  "       example: --indexed-otn Word\n");
	addOption("--indexed-feature", "--indexed-feature", true,
		  "",
		  "ERROR: --indexed-feature must have a feature name as its argument.\n"
		  "       example: --indexed-feature surface\n");
	addOption("--nodb", "--nodb", false);
	addOption("--no-vacuum", "--no-vacuum", false);

	addOption("-d", "--dbname", true, 
		  "emdf",
		  "ERROR: -d and --dbname must have a database-name as their argument.\n"
		  "       example: -d book1\n");
	addOption("-o", "--output", true, 
		  "",
		  "ERROR: -o and --output must have a filename as their argument.\n"
		  "       example: -o indexcreation.mql\n");
	addOption("-fe", "--fts-engine", true, 
		  "1",
		  "ERROR: -fe and --fts-engine must have a number as their argument.\n"
		  "       example: -fe 2\n");

	// Parse arguments
	std::list<std::string> surplus_arguments;

	std::string error_message;
	if (!parseArguments(argc, argv, error_message, surplus_arguments)) {
		print_usage(std::cerr);
		std::cerr << error_message << std::endl;
		return 1;
	} else {
		bool bShowVersion;
		bool bShowHelp;
		std::string error_message;
		if (!getStandardArguments(bShowVersion, bShowHelp,
					  source_hostname,
					  source_user,
					  source_password,
					  source_backend_kind,
					  charset,
					  error_message)) {
			print_usage(std::cerr);
			std::cerr << error_message << std::endl;
			return 1;
		}

		// -V | --version
		if (bShowVersion) {
			show_version(std::cout);
			return 0;
		}

		// --help
		if (bShowHelp) {
			print_usage(std::cout);
			return 0;
		}

		// stylesheet name
		if (getArgumentPresent("-s")) {
			getArgumentValue("-s", stylesheet_name);
		} else {
			stylesheet_name = "";
		}

		// stylesheet filename
		if (getArgumentPresent("-sf")) {
			getArgumentValue("-sf", stylesheet_file_name);
		} else {
			stylesheet_file_name = "";
		}

		if ((stylesheet_name != "" &&
		     stylesheet_file_name == "")
		    || (stylesheet_name == "" &&
			stylesheet_file_name != "")) {
			print_usage(std::cerr);
			std::cerr << "The -s <stylesheet> and -sf <stylesheet-filename> options\nmust be given together!" << std::endl;
			return 1;
		}

		// source database name
		if (getArgumentPresent("-d")) {
			getArgumentValue("-d", source_db_name);
		} else {
			print_usage(std::cerr);
			std::cerr << "The -d <database-name> option is required!" << std::endl;
			return 1;
		}

		// target database name
		if (getArgumentPresent("-td")) {
			getArgumentValue("-td", target_db_name);
		} else {
			target_db_name = source_db_name;
		}

		// target database hostname
		if (getArgumentPresent("-th")) {
			getArgumentValue("-th", target_hostname);
		} else {
			target_hostname = source_hostname;
		}

		// target database user
		if (getArgumentPresent("-tu")) {
			getArgumentValue("-tu", target_user);
		} else {
			target_user = source_user;
		}

		// target database password
		if (getArgumentPresent("-tp")) {
			getArgumentValue("-tp", target_password);
		} else {
			target_password = source_password;
		}

		// target database backend
		if (getArgumentPresent("-tb")) {
			std::string target_backend_name;
			if (!string2backend_kind(target_backend_name, target_backend_kind)) {
				print_usage(std::cerr);
				std::cerr << "Unknown target backend kind '" << target_backend_name << "'" << std::endl;
				return 1;
			}
		} else {
			target_backend_kind = source_backend_kind;
		}

		// bookcase-otn
		if (getArgumentPresent("--bookcase-otn")) {
			getArgumentValue("--bookcase-otn", bookcase_otn);
		} else {
			print_usage(std::cerr);
			std::cerr << "The --bookcase-otn <bookcase-otn> option is required!" << std::endl;
			return 1;
		}

		// indexed-otn
		if (getArgumentPresent("--indexed-otn")) {
			getArgumentValue("--indexed-otn", indexed_otn);
		} else {
			print_usage(std::cerr);
			std::cerr << "The --indexed-otn <indexed-otn> option is required!" << std::endl;
			return 1;
		}

		// indexed-feature
		if (getArgumentPresent("--indexed-feature")) {
			getArgumentValue("--indexed-feature", indexed_feature);
		} else {
			print_usage(std::cerr);
			std::cerr << "The --indexed-feature <indexed-feature> option is required!" << std::endl;
			return 1;
		}

		// output_filename
		if (getArgumentPresent("-o")) {
			getArgumentValue("-o", output_filename);
		} else {
			output_filename = "";
		}

		// fts engine version
		if (getArgumentPresent("-fe")) {
			std::string fts_engine_str;
			getArgumentValue("-fe", fts_engine_str);
			fts_engine_version = (int) string2long(fts_engine_str);
		} else {
			fts_engine_version = 1;
		}

		if (getArgumentPresent("--nodb")) {
			bDontUseTargetDB = true;
		}  else {
			bDontUseTargetDB = false;
		}

		if (getArgumentPresent("--no-vacuum")) {
			bEmitVacuum = false;
		} else {
			bEmitVacuum = true;
		}

		if (surplus_arguments.size() > 0) {
			if (stylesheet_name != "") {
				print_usage(std::cerr);
				std::cerr << "You cannot both give the -s and -sf options, and have filters on the\ncommandline." << std::endl;
				return 1;
			} else {
				std::list<std::string>::const_iterator ci = 
					surplus_arguments.begin();
				while (ci != surplus_arguments.end()) {
					filter_string_list.addStringBack(*ci);
					++ci;
				}
			}
		} 
	}

	if (!(fts_engine_version == 1 
	      || fts_engine_version == 2)) {
		std::string error_message = "ERROR: Requested FTS engine version " + long2string(fts_engine_version) + " is not available.\n"
			+ "Please use either 1 or 2.";
		print_usage(std::cerr);
		std::cerr << error_message << std::endl;
		return 1;
	} 

	// Make connection

	EmdrosEnv *pSourceEE = new EmdrosEnv(output_kind, 
					     charset,
					     source_hostname,
					     source_user, source_password,
					     source_db_name,
					     source_backend_kind);

	EmdrosEnv *pTargetEE = 0;
	if (bDontUseTargetDB) {
		pTargetEE = 0;
	} else {
		if (source_db_name != target_db_name
		    || source_hostname != target_hostname
		    || source_user != target_user
		    || source_password != target_password
		    || source_backend_kind != target_backend_kind) {
			pTargetEE = new EmdrosEnv(output_kind, 
						  charset,
						  target_hostname,
						  target_user, target_password,
						  target_db_name,
						  target_backend_kind);
		} else {
			pTargetEE = pSourceEE;
		}
	}

	// Zero-fill passwords
	zeroFillString(source_password);
	zeroFillString(target_password);

	// Check that we connected to source
	if (!(pSourceEE->connectionOk())) {
		emdrosMessageConnectionNotOK(&std::cerr, source_backend_kind);
		if (pSourceEE == pTargetEE) {
			delete pSourceEE;
		} else {
			delete pSourceEE;
			delete pTargetEE;
		}
		return 2;
	}

	// Check that we connected to target
	if (pTargetEE && !pTargetEE->connectionOk()) {
		emdrosMessageConnectionNotOK(&std::cerr, target_backend_kind);
		if (pSourceEE == pTargetEE) {
			delete pSourceEE;
		} else {
			delete pSourceEE;
			delete pTargetEE;
		}
		return 2;
	}

	int nResult = 3;
	try {
		nResult = exec_indexer(pSourceEE, pTargetEE, output_filename, bookcase_otn, indexed_otn, indexed_feature, stylesheet_name, stylesheet_file_name, filter_string_list, bEmitVacuum, fts_engine_version);
	} catch (EMdFNULLValueException& e) {
		std::cerr << "ERROR: EMdFNULLValueException (Database error)..." << std::endl;
		std::cerr << pSourceEE->getDBError() << std::endl; 
		std::cerr << pSourceEE->getCompilerError() << std::endl;
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBDBError& e) {
		std::cerr << "ERROR: EMdFDBDBError (Database error)..." << std::endl;
		std::cerr << pSourceEE->getDBError() << std::endl; 
		std::cerr << pSourceEE->getCompilerError() << std::endl;
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBException& e) {
		std::cerr << "ERROR: EMdFDBException (Database error)..." << std::endl;
		std::cerr << pSourceEE->getDBError() << std::endl; 
		std::cerr << pSourceEE->getCompilerError() << std::endl;
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableColumnException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableRowException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (BadMonadsException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (WrongCharacterSetException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFOutputException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EmdrosException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (...) {
		std::cerr << "Unknown exception occurred.  Program aborted." << std::endl;
	} 

	// Clean up
	if (pSourceEE == pTargetEE) {
		delete pSourceEE;
	} else {
		delete pSourceEE;
		delete pTargetEE;
	}
	return nResult;
}


